### PR TITLE
fix: cjs import in vite macro

### LINKED ISSUE #512: The macro fix for vite added in #386 does not work
Hi there, I'm trying to use the macro for `vite` using the instructions from the README.

I get this error when I run `yarn dev`:

```
yarn run v1.22.19
$ vite
failed to load config from /Users/beau/p/test/vite.config.ts
error when starting dev server:
file:///Users/beau/p/test/node_modules/valtio/esm/macro/vite.mjs:4
import { MacroError } from 'babel-plugin-macros';
         ^^^^^^^^^^
SyntaxError: Named export 'MacroError' not found. The requested module 'babel-plugin-macros' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'babel-plugin-macros';
const { MacroError } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:128:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:194:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:409:24)
    at async loadConfigFromBundledFile (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-71eb12cb.js:62893:21)
    at async loadConfigFromFile (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-71eb12cb.js:62780:28)
    at async resolveConfig (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-71eb12cb.js:62398:28)
    at async createServer (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-71eb12cb.js:59018:20)
    at async CAC.<anonymous> (file:///Users/beau/p/test/node_modules/vite/dist/node/cli.js:699:24)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

This seems to be an issue with mixing ESM (which vite converts everything to, if I understand correctly how it works) and CJS.

I'm using node v18.4.0 and the most recent vite, valtio, react, and typescript as of today.

Here's my `vite.config.js`:

```
import { defineConfig } from 'vite';
import macro from 'valtio/macro/vite';
import react from '@vitejs/plugin-react';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    macro,

    react({
      babel: {
        plugins: ['@emotion/babel-plugin'],
      },

      jsxImportSource: '@emotion/react',
    }),
  ],
});
```

### PR DESCRIPTION
**Pull request by Aslemammad**
Resolves #512 
This is a trial to fix the cjs import issue with converting the normal import to node cjs compatible import!

### CODE REVIEW

**Comment by dai-shi** [comment]
btw, do you want to support vite 3?

**Comment by Aslemammad** [comment]
@dai-shi Is there any problem with Vite 3, let me know if you see any issue there.

**Comment by dai-shi** [comment]
I'm not sure if there's a real issue, but we stick with v2.8.6,

https://github.com/pmndrs/valtio/blob/7c85c1413992a02e7a65e698b803fdef500d008e/package.json#L198

because we need [a patch](https://github.com/pmndrs/valtio/blob/7c85c1413992a02e7a65e698b803fdef500d008e/patches/vite%2B2.8.6.patch).

**Comment by Aslemammad** [comment]
@dai-shi I don't think there's a real problem there, let's keep it as is!

File: package.json [hunk 1]
```diff
@@ -53,7 +53,6 @@
   ],
   "sideEffects": false,
   "scripts": {
-    "postinstall": "patch-package",
```
Comment (dai-shi) [inline package.json:56]: Please also remove it from `devDependencies`.

Comment (Aslemammad) [reply package.json:56]: It is removed indeed!

File: src/macro/vite.ts [hunk 2]
```diff
@@ -1,22 +1,23 @@
...
   .withSignature(`<T extends object>(proxyObject: T): void`)
   .withHandler((ctx) => {
     const { path, args } = ctx
-    const hook = addNamed(path, 'useSnapshot', 'valtio')
+    const hook = babelModuleImports.addNamed(path, 'useSnapshot', 'valtio')
     const proxy = args[0]?.node
 
-    if (!t.isIdentifier(proxy)) throw new MacroError('no proxy object')
+    if (!t.isIdentifier(proxy))
+      throw new babelMacro.MacroError('no proxy object')
```
Comment (dai-shi) [inline src/macro/vite.ts:20]:
```suggestion
    if (!t.isIdentifier(proxy)) {
      throw new babelMacro.MacroError('no proxy object')
    }
```

**Review by dai-shi — APPROVED** [review]
LGTM
Do we wait for a response in #512?

**Comment by Aslemammad** [comment]
@dai-shi I believe we should do, it'd be better!

**Comment by beaugunderson** [comment]
progress, but now fails on the import from `slemammad-vite-plugin-macro`:

```
import { defineMacro, createMacroPlugin, defineMacroProvider } from 'aslemammad-vite-plugin-macro';
         ^^^^^^^^^^^
SyntaxError: Named export 'defineMacro' not found. The requested module 'aslemammad-vite-plugin-macro' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'aslemammad-vite-plugin-macro';
const { defineMacro, createMacroPlugin, defineMacroProvider } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:123:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:189:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:541:24)
    at async loadConfigFromBundledFile (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-5cb728cb.js:62986:21)
    at async loadConfigFromFile (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-5cb728cb.js:62872:28)
    at async resolveConfig (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-5cb728cb.js:62486:28)
    at async createServer (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-5cb728cb.js:59089:20)
    at async CAC.<anonymous> (file:///Users/beau/p/test/node_modules/vite/dist/node/cli.js:699:24)
```

to test this I used this build:

```
yarn add valtio@https://pkg.csb.dev/pmndrs/valtio/commit/1a0d5e66/valtio/_pkg.tgz
```

**Comment by Aslemammad** [comment]
@beaugunderson Feel free to try it now!

**Comment by beaugunderson** [comment]
hmm, still no go:

```
$ vite
failed to load config from /Users/beau/p/test/vite.config.ts
error when starting dev server:
TypeError: plugin.defineMacro is not a function
    at file:///Users/beau/p/test/node_modules/valtio/esm/macro/vite.mjs:6:28
    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:541:24)
    at async loadConfigFromBundledFile (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-5cb728cb.js:62986:21)
    at async loadConfigFromFile (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-5cb728cb.js:62872:28)
    at async resolveConfig (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-5cb728cb.js:62486:28)
    at async createServer (file:///Users/beau/p/test/node_modules/vite/dist/node/chunks/dep-5cb728cb.js:59089:20)
    at async CAC.<anonymous> (file:///Users/beau/p/test/node_modules/vite/dist/node/cli.js:699:24)
```

**Comment by Aslemammad** [comment]
@beaugunderson Could you share your node version?

**Comment by beaugunderson** [comment]
@Aslemammad yes, it's `v18.7.0`, using vite `3.0.5`

**Comment by beaugunderson** [comment]
just tried via `nvm` with 14, 15, 16, 17, all fail with the same issue

**Comment by beaugunderson** [comment]
here is a minimal codesandbox that shows the failure:

https://codesandbox.io/s/vite-react-ts-forked-gm49m0?file=/package.json

I appreciate you work on getting this fixed btw!

**Comment by Aslemammad** [comment]
@beaugunderson Thank you mate, I'm trying to fix it! hope I can deliver it soon!

**Comment by beaugunderson** [comment]
this might actually be a little better; vite links to stackblitz for their various templates; this is the `react-ts` template with the only change being following the README steps from valtio to add the patch, boots right to the failure:

https://stackblitz.com/edit/vitejs-vite-qsaxht

**Comment by beaugunderson** [comment]
note that in the `react-ts` template `"type": "module"` is in `package.json`, which I think is what's causing the issue... just not sure the right way to fix :)

**Comment by Aslemammad** [comment]
@beaugunderson It now works on my system with your repro, feel free to try it out.

**Pull request merged**